clang/lib/DPCT/CUBAPIMigration.cpp

```diff
@@ -172,11 +178,133 @@ void CubMemberCallRule::registerMatcher(ast_matchers::MatchFinder &MF) {
       this);
 }
 
+static std::pair<const VarDecl *, TypeLoc>
+getTempstorageVarAndValueTypeLoc(const CXXMemberCallExpr *MC) {
+  Expr *Obj = MC->getImplicitObjectArgument();
+  const VarDecl *TempStorage = nullptr;
+
+  auto FindTempStorageVarInCtor = [&](const Expr *E) -> const VarDecl * {
+    if (auto *Ctor = dyn_cast<CXXConstructExpr>(E)) {
+      if (auto *DRE = dyn_cast<DeclRefExpr>(Ctor->getArg(0))) {
+        if (auto *VD = dyn_cast<VarDecl>(DRE->getDecl())) {
+          if (VD->hasAttr<CUDASharedAttr>() && isCubVar(VD)) {
+            return VD;
+          }
+        }
+      }
+    }
+    return nullptr;
+  };
+
+  auto HandleTypeLoc = [&](TypeLoc Loc) -> TypeLoc {
+    if (Loc.isNull())
+      return Loc;
+    while (true) {
+      switch (Loc.getTypeLocClass()) {
+      case TypeLoc::Elaborated:
+        Loc = Loc.getNextTypeLoc();
+        break;
+      case TypeLoc::Typedef: {
+        auto NewLoc = Loc.castAs<TypedefTypeLoc>();
+        Loc = NewLoc.getTypedefNameDecl()->getTypeSourceInfo()->getTypeLoc();
+        break;
+      }
+      case TypeLoc::TemplateSpecialization: {
+        auto NewLoc = Loc.getAs<TemplateSpecializationTypeLoc>();
+        return NewLoc.getArgLocInfo(0).getAsTypeSourceInfo()->getTypeLoc();
+        break;
+      }
+      default:
+        return Loc;
+      }
+    }
+  };
+
+  TypeLoc DataTypeLoc;
+  if (const auto *MTE = dyn_cast<MaterializeTemporaryExpr>(Obj)) {
+    if (auto *TOE = dyn_cast<CXXTemporaryObjectExpr>(MTE->getSubExpr())) {
+      DataTypeLoc = HandleTypeLoc(TOE->getTypeSourceInfo()->getTypeLoc());
```

Comment:
The name `HandleTypeLoc` is not clear, should be `getxxx`, like: `GetTargetTypeLoc`?